README.md
# BFD UI

[![build status](https://img.shields.io/travis/baifendian/bfd-ui.svg)](https://travis-ci.org/baifendian/bfd-ui)
[![npm package](https://img.shields.io/npm/v/bfd-ui.svg)](https://www.npmjs.org/package/bfd-ui)
[![NPM downloads](http://img.shields.io/npm/dm/bfd-ui.svg)](https://npmjs.org/package/bfd-ui)

企业级 React 组件库 [http://ui.baifendian.com](http://ui.baifendian.com)

## 安装

```sh
npm install bfd-ui --save
```

## 使用

BFD UI 抛弃传统的资源加载方式，基于 [webpack](https://webpack.github.io/) 打包，资源种类多种多样，会涉及一些 webpack 的配置，如下：

```js
{
  module: {
    loaders: [{
      test: /\\.(eot|woff|woff2|ttf|svg|png|jpg)(\\?v=[\\d\\.]+)?$/,
      loader: 'file?name=files/[hash].[ext]'
    }, {
      test: /\\.css$/,
      loader: 'style!css'
    }]
  },
  resolve: {
    alias: {
      bfd: 'bfd-ui/lib'
    }
  }
}
```
webpack 配置完成后，即可在代码中使用组件，以 [DatePciker](http://ui.baifendian.com/components/DatePicker) 为例：
```js
import React, { Component } from 'react'
import DatePicker from 'bfd/DatePicker'

class App extends Component {

  handleSelect(date) {
    console.log(date)
  }

  render() {
    return <DatePicker onSelect={this.handleSelect} />
  }
}
```
> 强烈建议基于[脚手架](https://github.com/baifendian/create-bfd-app)开发，省去繁琐的环境配置、项目基本结构、模板以及开发和部署相关工作


## 组件全局配置

覆盖或扩展 `defaultProps` 即可，以 [DatePciker](http://ui.baifendian.com/components/DatePicker) 为例：
```js
DatePicker.defaultProps = Object.assign(DatePicker.defaultProps || {}, {
  placeholder: 'Please select date'
})
```
命令式 API 模块 [message](http://ui.baifendian.com/components/message)、[confirm](http://ui.baifendian.com/components/confirm)、[xhr](http://ui.baifendian.com/components/xhr) 也支持全局配置，涉及 url 方式加载数据的组件以及 [Form](http://ui.baifendian.com/components/Form) 均依赖 xhr。详细配置请参考其各自文档

> 全局配置后，这些 API 会变成有状态的，即最终结果受配置影响，所以尽量一次性配置并向其它开发者说明


## 受控属性与不受控属性

组件支持不受控（`defaultXXX`）的使用方式，关于为什么区分受控 / 不受控请参考 [Controlled / Uncontrolled Components](https://facebook.github.io/react/docs/forms.html#controlled-components)


## 浏览器支持

Chrome、Firefox、Safari、IE9+


## 开发者说明

#### 开发环境安装

```sh
git clone https://github.com/yourName/bfd-ui.git

# DEMO 站集成了脚手架项目的说明，所以需要依赖这个项目的一些资源，与 bfd-ui clone 在同一目录下即可
git clone https://github.com/baifendian/create-bfd-app.git

cd bfd-ui

npm install

npm start
```
查看: http://localhost:4001


#### 开发规范

- 向下兼容
- 单元测试：组件根目录下创建 `__tests__` 文件夹
- 代码规范参考 [airbnb react](https://github.com/airbnb/javascript/tree/master/react)


#### 编写一个新组件
```sh
npm run create MyComponent
```
查看: http://localhost:4001/components/MyComponent


## 更新日志

[CHANGELOG](http://ui.baifendian.com/changelog)


## LICENSE

[BSD License](https://github.com/baifendian/bfd-ui/blob/master/LICENSE)


CHANGELOG.md
## v1.5.10

*2017-03-17*

- xhr 参数中增加async参数用于配置是否异步，同时type为get时支持自动将option.data中的参数拼接到url中（get模式下option.data注意不要使用object类型的值）
- form 增加init方法，用于重置form数据同时不触发rule检查。
- searchinput 修复修改value字段不能正常的显示值得问题。
- TreeSelect 修复修改value字段不能正常显示选中值得问题。


## v1.5.9

*2017-01-20*

- 修改了默认主色系，文档站首页样式升级
- 抽离了 css 公共值，统一管理


## v1.5.7

*2017-01-13*

- SearchInput 新增可控属性 value
- Modal 渲染到 body 节点内；样式及动画性能优化；修复嵌套 Modal 的样式问题；lock 模式添加 shake 效果
- 去除 Tabs 初始化时的动画
- 因 [pace](http://github.hubspot.com/pace/docs/welcome/) 加载进度不够精确，文档站页面切换进度条升级到 [nprocess](http://ricostacruz.com/nprogress/)，与 react-router 配合使用

>- Modal 渲染到 body 节点内，节点父子关系: `body > div > div.bfd-modal`
- v1.5.7 起，依赖的 react-router 升级到 2.x，项目中的 react-router 也需要升级到 2.x。 react-router 的兼容问题请参照 [react-router 2.x 升级日志](https://github.com/ReactTraining/react-router/blob/master/upgrade-guides/v2.0.0.md#changes-to-thiscontext)


## v1.5.6

*2017-01-06*

- FixedTable 新增属性 noDataContent
- 部分组件样式调整
- Dropdown DOM 结构调整

> Dropdown DOM 结构调整，Dropdown 不产生实际 DOM 节点，className 等属性请转移到 DropdownToggle

调整前
```markup
<div class="bfd-dropdown">
  <div class="bfd-dropdown__toggle bfd-popover__trigger">
    ...
  </div>
</div>
```
调整后
```markup
<div class="bfd-dropdown-toggle bfd-popover__trigger">
  ...
</div>
```


## v1.5.5

*2016-12-30*

- Tooltip 添加 `white-space: pre-wrap` 特性
- Dropdown, Tooltip, Select 等“浮层”组件支持任意嵌套
- Modal 新增属性 size


## v1.5.0

*2016-12-26*

- 新增组件 Slider2, SearchInput2
- 修正了部分组件样式
- README、GUIDE、文档字体更新


## v1.4.1

*2016-12-09*

- 调整了部分组件的动画效果
- 处理了 Popover 组件内容溢出的问题
- Input, ClearableInput value 支持 `number` 类型
- 处理了 AutoComplete 文字溢出的问题
- messgae: 支持消息队列；支持手动关闭；调整了样式
- 修复 DataTable column render index 参数错误的问题
- confirm 支持更多的配置项，新增全局配置 `config`、关闭 `close`


## v1.4.0

*2016-12-02*

> `v1.4.0` 起，DropdownMenu 渲染到 body 节点内，相关组件 Select, MultipleSelect 展开层 DOM 位置均改变，自定义的样式可能因 DOM 结构的改变而失效

- Select onChange 新增参数 item
- DropdownMenu 渲染至 body 节点内，相关组件 Select, MultipleSelect 展开层 DOM 位置均改变
- Dropdown 新增属性 aligned，实现 DropdownToggle 与 DropdownMenu 宽度相同
- DropdownMenu 新增属性 direction, align，支持方向控制及对齐方式
- Tooltip, TextOverflow 新增属性 align，支持对齐方式
- 修复 Percentage css 宽度无效的问题

另外，新增大量属性支持自定义组件内部部分内容的显示，可用于国际化等需求，新增相关组件及属性如下

- `Select`: searchPlaceholder, noOptionsContent, noMatchingContent
- `CheckboxGroup`: toggleAllContent
- `Paging`: auxiliaryRender
- `MultipleSelect`: noOptionsContent, toggleAllContent
- `DataTable`: noDataContent
- `DataTable2`: noDataContent
- `DatePicker/DateRange`: placeholder, captionRender, weekDayNames
- `Transfer`: searchPlaceholder
- `Tree`: noChildrenContent


全局配置方式，以 Select placeholder 为例：
```js
import { Select } from 'bfd/Select'

Object.assign(Select.defaultProps, {
  placeholder: 'Please select'
})
```


## v1.3.4

*2016-11-22*

- DataTable onRowClick 新增 index、event 参数
- TreeSelect 新增属性 dataFilter、disabled
- DataTable2 新增属性 rowRender，自定义行渲染；无数据显示处理；
- 修复 Select 无匹配时样式问题
- ButtonGroup 新增属性 value，修改为 inline-block 模式
- Fetch 新增 reload 接口，可强制重新加载
- 修复 TreeSelect value 为0时的问题
- Select、TreeSelect 新增属性 minWidth
- 修复 Upload 无法重复上传同一文件的问题 [#6](https://github.com/baifendian/bfd-ui/issues/6)
- Tooltip、TextOverflow 支持 className、style 等原生属性 [#7](https://github.com/baifendian/bfd-ui/issues/7)
- 修复 FormTextarea 不兼容 null 的问题
- Form 可不指定 data/defaultData；新增属性 sendDataFormatter，实现数据发送前的处理
- 修正了部分组件的样式


## v1.3.0

*2016-11-11* 😓

- 新增组件 DataTable2
- 修复 xhr data 共享传递的问题
- 修复 FormItem multiple 属性不可用的问题 [#4](https://github.com/baifendian/bfd-ui/issues/4)


## v1.2.1

*2016-11-04*

- MultipleSelect 全选置顶
- Tooltip、TextOverflow 新增属性 direction，支持自定义方向 [#3](https://github.com/baifendian/bfd-ui/issues/3)
- Tooltip 新增属性 triggerMode，支持自定义触发方式 [#3](https://github.com/baifendian/bfd-ui/issues/3)
- Editable 切换到编辑状态后自动选中；修复取消编辑后值未重置的问题
- TreeSelect 新增属性 getIcon、getUrl
- 修复 Tree 数据更新后选中节点的问题
- DataTable 列配置 render 方法增加列索引参数；新增属性 hideGo


## v1.2.0

*2016-10-21*

- 新增组件 Tooltip
- Percentage 新增属性 width 及样式升级
- DropdownMenu、DatePicker 样式升级
- 修复 NavItem blank 模式下 URL 变化的问题
- 修复 Checkbox、Radio 不支持 onMouseDown 等事件的问题
- 优化 Modal 性能，关闭状态下不再渲染
- 优化 Tabs 性能，重新渲染时隐藏的 TabPanel 不再渲染；TabPanel 增加淡入效果
- 优化 TextOverflow 交互效果并自适应显示位置
- 修复 Slider 宽度自适应问题


## v1.1.1

*2016-10-13*

- Tree、TreeSelect 新增属性 shouldNodeSelectable、shouldNodeCheckable
- Dropdown 自适应向上或向下展开
- 修复 ClearableInput、FormInput 额外占用 tabindex 的问题；修复 Select、Checkbox、Radio 等组件不响应 键盘 tab 切换的问题
- 修复 TextOverflow 子节点嵌套时不显示的问题
- Checkbox 新增属性 indeterminate，实现半勾选状态，并同步 Tree、CheckboxGroup
- Fetch 新增属性 spinnerHeight，控制加载动画的尺寸；修复异步回调时组件不存在的问题
- SearchInput 支持回车触发查询


## v1.1.0

*2016-09-29*

- 新增组件 TreeSelect
- Tree 新增属性 onSelect、checkable、onCheck；修复字体过大时样式问题
- MultipleSelect 新增属性 placeholder
- 修复 Form 中使用 DatePicker 选中日期后触发 submit 的问题
- 修复 Fetch 组件容器高度不随内容变化而变化的问题
- 修复 FixedTable 数据二次加载问题
- 修复 Upload 同一文件二次上传问题
- 修复 DataTable 单字节字符过多不自动换行问题


## v1.0.0

*2016-09-23*

- 第一个正式公开版本
- 修复 TextOverflow IE 下兼容问题


## v0.10.0

*2016-09-18*

> 自 `v0.10.0` 起新发布的版本，less 会编译成 css，即去除了编译 less 相关的依赖，需要的话请单独安装

- 新增组件 Spinner、Panel、FixedTable
- FormSubmit 新增 toggleProcess 接口
- 修复 Checkbox、Radio click 事件冒泡两次的问题
- 修复 react 15.x 环境下未知属性错误提醒的问题
- 预编译 less，不再依赖项目的 webpack less 相关配置
- NavItem 新增 blank 属性，实现新窗口打开
- DataTable 支持单个排序，支持复选框跨页选中记忆
- 修复 Paging 页码省略号显示问题，跳转问题
- Upload 新增 onUploading 上传进行中事件


## v0.9.1

*2016-09-09*

- 新增组件 Paging
- Dropdown 展开时添加三角标志；DropdownMenu 新增 right 属性，实现右对齐
- 修复 Select 搜索框宽度未自适应的问题
- Form 新增 FormSubmit 子组件；更新验证失败后的样式；新增 onSubmit 属性，自定义提交后的行为；优化验证方式，数据更新后自动验证
- xhr() 返回当前 XMLHttpRequest 对象; 新增全局配置 getUrl，实现动态 url
- 修复 Radio 无 children 时不可用的问题
- Fetch 新增 defaultHeight 属性，定义初始化高度
- 优化 Select 无匹配项时的显示
- Select、MultipleSelect 新增 dataFilter 属性，处理 url 模式的数据格式


## v0.8.0

*2016-09-05*

> 自 `v0.8.0` 起，去除了 bootstrap 的依赖，如果项目代码用到了 bootstrap 的特性，升级后请单独引入 `bfd-bootstrap` 模块

- Select、MultipleSelect 新增 data 属性
- Modal 新增 onToggle 属性
- CheckboxGroup、Form 新增 defaultSelects 属性
- Switch 新增 defaultOn 属性
- Row 新增 fluid 属性，流式布局
- message.danger 修改为自动关闭
- 导航组件新增 IndexNavItem，与 IndexRoute 对应
- 修复 Tree 动态加载数据父级重置后无法更新的问题

site/server.js
var express = require('express')
var http = require('http')
var path = require('path')
var bodyParser = require('body-parser');
var multiparty = require('multiparty');
var compression = require('compression')
var fs = require('fs')
var domain = require('domain')

var app = express()
app.use(compression())
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({
  extended: true
}));
app.use(express.static(path.join(__dirname, './')))

// Error handle
app.use(function(req, res, next) {
  var reqDomain = domain.create()
  var i = 0
  reqDomain.on('error', function(err) {
    res.send(500, '<pre>' + err.stack + '</pre>')
  })
  reqDomain.run(next)
})

if (process.env.NODE_ENV !== 'production') {
  var webpack = require('webpack')
  var webpackDevMiddleware = require('webpack-dev-middleware')
  var WebpackConfig = require('./webpack.config')
  app.use(webpackDevMiddleware(webpack(WebpackConfig), {
    publicPath: '/build/',
    stats: {
      colors: true,
      profile: true
    }
  }))
}

app.post('/api/form', function(req, res) {
  res.json({
    code: 200,
    data: {}
  })
})

function clone(obj) {
  var o;
  if (typeof obj == "object") {
    if (obj === null) {
        o = null;
    } else {
        if (obj instanceof Array) {
          o = [];
          for (var i = 0, len = obj.length; i < len; i++) {
              o.push(clone(obj[i]));
          }
        } else {
            o = {};
            for (var j in obj) {
                o[j] = clone(obj[j]);
            }
        }
    }
  } else {
      o = obj;
  }
  return o;
}

function getTableDate(page, size) {
  var data = [
    {id: 1, name: '张三', age: 28, gender: 'male', country: '中国', area: '北京', regdate: '2016-03-01' },
    {id: 2, name: '李四', age: 25, gender: 'female', country: '中国', area: '杭州', regdate: '2016-04-11' },
    {id: 3, name: '王五', age: 43, gender: 'male', country: '中国', area: '沈阳', regdate: '2016-05-06' },
    {id: 4, name: '赵某某', age: 30, gender: 'female', country: '中国', area: '上海', regdate: '2016-03-09' },
    {id: 5, name: '钱某某', age: 39, gender: 'male', country: '中国', area: '深圳', regdate: '2015-11-11' },
    {id: 6, name: '孙某某', age: 50, gender: 'male', country: '中国', area: '石家庄', regdate: '2016-06-01' },
    {id: 7, name: '周某某', age: 21, gender: 'female', country: '中国', area: '西安', regdate: '2016-08-13' },
    {id: 8, name: '吴某某', age: 19, gender: 'female', country: '中国', area: '天津', regdate: '2016-02-22' },
    {id: 9, name: '郑某某', age: 51, gender: 'male', country: '中国', area: '武汉', regdate: '2016-01-18' },
    {id: 10, name: '冯某某', age: 24, gender: 'male', country: '中国', area: '广州', regdate: '2016-09-20' }
  ]

  if(page <= 1) {
    return data
  } else {
    var index = page * size
    var newData = []
    for(var i=0; i<size && i<10; i++) {
      var item = clone(data[i])
      item.id = item.id + index;
      item.name += item.id;
      newData.push(item)
    }
    return newData
  }
}

app.get('/api/table', function(req, res) {
  var currentPage = req.query.currentPage || 1
  var pageSize = req.query.pageSize || 10
  var data = getTableDate(currentPage, pageSize)

  res.json({
    code: 200,
    totalList: data,
    currentPage: 1,
    totalPageNum: 300
  })

})

app.post('/upload.do', function(req, res) {
  var form = new multiparty.Form({uploadDir: './upload'});
  form.on('error', function(err) {
    console.log('Error parsing form: ' + err.stack);
  });
  form.parse(req, function (err, fields, files){
    var filesTmp = JSON.stringify(files,null,2);
    try{
      if (err){
        console.log('parse error: ' + err);
        res.send("写文件操作失败。");
      }else {
        res.send("文件上传成功");
        console.log('parse files: ' + filesTmp);
        var fileNameArr = Object.keys(files);
        var firstFilename = fileNameArr[0];
        var fileDataArr = files[firstFilename];
        var fileData = fileDataArr[0];
        var uploadedPath = fileData.path;
        var dstPath = './upload/' + fileData.originalFilename;
        fs.rename(uploadedPath, dstPath, function(err) {
          if (err){
            console.log("重命名文件错误："+ err);
          } else {
            console.log("重命名文件成功。");
          }
        });
      }
    } catch(e) {}
  });
  res.json({
    code: 200,
    data: {
      id: new Date().getTime()
    }
  })
})

app.get('*', function(req, res) {
  res.sendFile(path.join(__dirname, 'index.html'))
})

var port = process.env.PORT || 4001

app.listen(port, function() {
  console.log('Server listening on http://localhost:' + port + ', Ctrl+C to stop')
})
